core/tracestack.lua

```diff
@@ -0,0 +1,214 @@
+        line = line or content.line,
+        column = column or content.col,
+        options = options or content.attr or {}
+      })
+  end,
+
+  -- Push a text that is going to get typeset on to the stack to record the execution trace for debugging.
+  -- Must be popped with `pop(returnOfPush)`.
+  pushText = function(self, text, line, column, file)
+    return self:_pushFrame({
+        text = text,
+        file = file,
+        line = line,
+        column = column
+      })
+  end,
+
+  -- Internal: Push complete frame
+  _pushFrame = function (self, frame)
+    frame.toLocationString = self._frameToLocationString
+    frame.typesetter = SILE.typesetter
+    -- Push the frame
+    if SU.debugging("commandStack") then
+      print(string.rep(" ", #self) .. "PUSH(" .. frame:toLocationString(false, true) .. ")")
+    end
+    self[#self + 1] = frame
+    self.afterFrame = nil
+    self._lastPushId = self._lastPushId + 1
+    frame.pushId = self._lastPushId
+    return self._lastPushId
+  end,
+
+  -- Pop previously pushed command from the stack.
+  -- Return value of `push` function must be provided as argument to check for balanced usage.
+  pop = function(self, pushId)
+    if type(pushId) ~= "number" then
+      SU.error("SILE.traceStack:pop's argument must be the result value of the corresponding push", true)
+    end
+    -- First verify that push/pop is balanced
+    local popped = self[#self]
+    if popped.pushId ~= pushId then
+      local message = "Unbalanced content push/pop"
+      local debug = SILE.traceback or SU.debugging("commandStack")
+      if debug then
+        message = message .. ". Expected " .. popped.pushId .. " - (" .. popped:toLocationString() .. "), got " .. pushId
+      end
+      SU.warn(message, debug)
+    else
+      -- Correctly balanced: pop the frame
+      self.afterFrame = popped
+      self[#self] = nil
+      if SU.debugging("commandStack") then
+        print(string.rep(" ", #self) .. "POP(" .. popped:toLocationString(false, true) .. ")")
+      end
+    end
+  end,
+
+  -- Returns frame stack represented by SILE.traceStack and self.afterFrame, but without frames whose typesetter
+  -- is different than what is given as a first parameter (nil = use current typesetter for filtering).
+  -- Most relevant object is in the last index.
+  filteredByTypesetter = function(self, typesetter)
+    typesetter = typesetter or SILE.typesetter
+    local result = {}
+    for i = 1, #self do
+      if self[i].typesetter == typesetter then
+        result[#result + 1] = self[i]
+      end
+    end
+    local afterFrame
+    if self.afterFrame and self.afterFrame.typesetter == typesetter then
+      afterFrame = self.afterFrame
+    end
+    return result, afterFrame
+  end,
+
+  -- Returns short string with most relevant location information for user messages.
+  locationInfo = function(self)
+    return self:_formatTraceHead(self.afterFrame) or SILE.currentlyProcessingFile or "<nowhere>"
```

Comment:
Why is this function always operating on the _last item_ popped from the stack? Wouldn't it be more relevant to operate on the top of the stack? As in the last thing pushed instead of the last thing popped?

Reply:
It is called with first parameter being the stack itself (notice the `:` instead of `.`), so it primarily does operate on the top of the stack.

Reply:
Yet the function it is calling doesn't name the variable as `self`, it names it something else which starts to get confusing. Basically you're passing it something it already has access to just to tip it off to use what it knows rather than ignore it. It's like _Inception_ ... don't worry I'm experimenting with ways around this.

Reply:
It is not called `self`, because it is not always called with `:`. In this case I am simply using the `:` as a syntactic shortcut to not have to write `self` twice.